### PR TITLE
Installation of PostgreSQL via Dockerfile

### PR DESCRIPTION
**Pull request by alextsaihi**
- Installation steps for PostgreSQL and dependencies is done

- Current issue with unknown cause/resolution: docker build is unable to call psql command after completing the installation of PostgreSQL
- still unable to create tables and load database from here

### CODE REVIEW

File: Dockerfile
Comment (yash-puligundla) [inline Dockerfile]: This is a duplicate of line 24. Could we remove this line?

File: cat
Comment (yash-puligundla) [inline cat:1]: Can we remove this empty file?

File: Makefile
Comment (yash-puligundla) [inline Makefile]: I think we should keep "clean-psql" as a part of `psql-db-refresh` to ensure that any existing databases are removed before creating a new one.

File: Dockerfile
Comment (yash-puligundla) [inline Dockerfile]: set the PATH variable so that psql can be accessed from inside the docker container
```suggestion
 export PATH="/usr/local/pgsql:${PATH}"
```

File: Dockerfile
Comment (yash-puligundla) [inline Dockerfile]: Add creation of PostgreSQL database, tables, and populate tables.

File: Dockerfile
Comment (yash-puligundla) [inline Dockerfile]: Keep the existing Dockerfile as is. I would move the PostgreSQL-related changes into a new Dockerfile named `Dockerfile-Postgres` so that we don't mix up the databases and also keep the existing docker image light-weight.

**Review by yash-puligundla — CHANGES_REQUESTED** [review]
Hi @alextsaihi I have added my comments in-line. Please let me know if you have any questions.

File: Dockerfile
Comment (alextsaihi) [reply Dockerfile]: These steps are in `RUN make psql-db-refresh`  line, but because I'm encountering a "psql not found" error even after adding psql to the Docker PATH, Jimmy said to just leave the ticket as is here.

**Comment by alextsaihi** [comment]
Hi @yash-puligundla thanks for the comments, also fixed some details you caught that I had overlooked. 

Just a question regarding the new Dockerfile-Postgres file. Would this function as a Dockerfile on its own, or just something to be copy/pasted into the Dockerfile in a later commit when the psql bug has been fixed? 
There are resources online for multiple Dockerfiles within one project, where each Dockerfile (named Dockerfile) were located in separate folders, but seemed like something different from what we are trying to achieve here.

**Comment by yash-puligundla** [comment]
> Hi @yash-puligundla thanks for the comments, also fixed some details you caught that I had overlooked.
> 
> Just a question regarding the new Dockerfile-Postgres file. Would this function as a Dockerfile on its own, or just something to be copy/pasted into the Dockerfile in a later commit when the psql bug has been fixed? There are resources online for multiple Dockerfiles within one project, where each Dockerfile (named Dockerfile) were located in separate folders, but seemed like something different from what we are trying to achieve here.

Hi @alextsaihi "Dockerfile-Postgres" is a dockerfile on its own in the same directory as the "Dockerfile". To build an image you can use "-f" flag to use the file. Example: 
`docker build -f Dockerfile-Postgres -t ${DOCKER_IMG} --build-arg VERSION=${DOCKER_TAG} .`

Let me know if this answered your question.

**Comment by yash-puligundla** [comment]
@alextsaihi It appears that there could be a misconception regarding the `Dockerfile-Postgres`. The intention is for the Docker image created using `Dockerfile-Postgres` to have identical functionality to the image produced by `Dockerfile`. The only difference is that the former uses Postgres for the database, whereas the latter uses SQLite. 

Therefore, the Dockerfile-Postgres should contain all the code present in the Dockerfile with the exception that the portions related to SQLite should be replaced with corresponding Postgres instructions. Please let me know if you have any further questions. Thank you!

**Comment by alextsaihi** [comment]
@yash-puligundla I added the updates on dockerfile-postgres to act as a standalone dockerfile available to be called via the -f flag

**Comment by yash-puligundla** [comment]
> @yash-puligundla I added the updates on dockerfile-postgres to act as a standalone dockerfile available to be called via the -f flag

Thank you @alextsaihi. Let's remove the SQLite installation from "dockerfile-postgres". For the current feature, the database must be created using PostgreSQL (not SQLite). I will hold off on approval of this Pull Request since we do not have a PostgreSQL database being created.

File: Dockerfile-Postgres
Comment (yash-puligundla) [inline Dockerfile-Postgres]: @alextsaihi when I tried this I had to use `export PATH="/usr/local/pgsql:${PATH}"` (pgsql and not psql) and it was able to recognize psql command after setting this PATH variable. I would check if that's the case here as well.

Comment (alextsaihi) [reply Dockerfile-Postgres]: I'll leave the changes to pgsql then, because neither of those variations worked on my end.

Comment (yash-puligundla) [reply Dockerfile-Postgres]: @alextsaihi 
Also, if you're not already familiar, both the Dockerfile-postgres and Dockerfile consists of three stages:

Stage 1: focuses on installing Postgres (SQLite in the case of Dockerfile) and building the database.

Stage 2: JAR file is generated.

Stage 3: Executing the application using the JAR file.

It's important to note that in stage 3, the database file (created in stage 1) is copied and used. However, note that the PostgreSQL executable isn't accessible within this stage. Hence, attempting to run PostgreSQL commands within stage 3 (and even stage 2) won't function **as per the design**. However PostgreSQL commands do work within stage 1 where you need it to create and load the db (I can confirm this based on my own testing).

Comment (alextsaihi) [reply Dockerfile-Postgres]: Thanks for the explanation, this does help clarify the process of Dockerfiles and Makefiles for me. 

However, the issue that I was having was from the line in Dockerfile that calls for psql refresh, which runs the psql commands from the Makefile that does db building and populating. When it tries to run this step, it tells me that psql is not found.

**Comment by alextsaihi** [comment]
hi @yash-puligundla, i've made some additions to this PR:
- docker compose for running postgresql and sk-drs
- separate dockerfile for postgres container

if the changes are good, i'll update the README before the merge

**Comment by yash-puligundla** [comment]
@alextsaihi, Before I proceed with the code review, could you please ensure that you have thoroughly tested the changes and confirmed that they adhere to the acceptance criteria as intended? Please provide the steps/commands and expected outcomes to test this pull request.

**Comment by alextsaihi** [comment]
Hi @yash-puligundla , the work is finally done! (i hope) 

steps to test:

FROM POSTGRES
1. `docker-compose up -d`
2. from postgres-1 container's terminal, `psql -U postgres`
3. `\c starter_kit_db` -> you should see "You are now connected to database "starter_kit_db" as user "postgres"."
4. `\dt`-> you should see 9 tables listed 
5. `\q` (to exit postgres terminal)

FROM DRS
from Postman, GET the following urls:
- `http://localhost:5000/ga4gh/drs/v1/service-info`
- `http://localhost:5000/ga4gh/drs/v1/objects/1a570e4e-2489-4218-9333-f65549495872`

CLOSING TEST
stop postgres-1 container and try a GET request again

**Comment by dashrath-chauhan** [comment]
Hi @alextsaihi, before we proceed ahead with merging code changes, we need to make sure that all the workflow tests gets executed without any errors. Github actions triggers the workflow execution everytime new code is pushed to the repository, so make sure that workflow gets executed without any errors and then proceed with opening a pull request.

File: Makefile
Comment (yash-puligundla) [inline Makefile]: since psql is not being installed in default docker and in local environment, I would remove it from here. In the current state, it will error with "command not found: psql" which is as expected

File: Dockerfile
Comment (yash-puligundla) [inline Dockerfile]: since there is no change, I would revert this

File: .gitignore
Comment (yash-puligundla) [inline .gitignore]: since there is no change, I would revert this

File: docker-compose-postgres.yml
Comment (yash-puligundla) [inline docker-compose-postgres.yml]: remove any unused and commented out code

File: README.md
Comment (yash-puligundla) [inline README.md:71]: Could you please include the expected outcomes in your explanation? For example, when running docker ps, which containers should be listed? What port is the PostgreSQL database available on, and can you briefly describe the connection process between the DRS Starter Kit Docker image and the PostgreSQL database using the configuration file? It would be helpful if you could mention which specific configuration fields are utilized to establish the connection with the PostgreSQL database.

File: src/test/resources/config/test-postgres.yml
Comment (yash-puligundla) [inline src/test/resources/config/test-postgres.yml:1]: since this file is not used for testing purposes, shouldn't we move it to the root folder. Consider naming it something like "config-drs-postgres.yml"

File: README.md
Comment (yash-puligundla) [inline README.md]: This should be a command
```suggestion
`docker compose -f docker-compose-postgres.yml up -d`
```

File: README.md
Comment (yash-puligundla) [reply README.md:71]: Please add information on how to shutdown the services that were created by running the above command.

**Review by yash-puligundla — CHANGES_REQUESTED** [review]
I have provided comments in-line, please let me know if you have questions.
Additionally, the github actions pipeline is failing. This is probably related to line 27 in Makefile. Please ensure that the changes in this PR do not break the github actions pipeline.

**Comment by yash-puligundla** [comment]
@alextsaihi Figured out why Github Actions is failing on your PR. It requires credentials that are stored as Github repo secrets. Since you are using a fork, your repo does not have these secret variables. Could you please add your feature branch to the main repository and then run Github Actions? Additionally, make sure you rebased your branch on develop. I see that the `at-postgres` branch is behind by a few commits which are crucial in a successful Github Actions run.

**Comment by alextsaihi** [comment]
Migrated to #83 , we can close this PR and discuss on the new one.

**Pull request closed without merging**
